src/ray/raylet/scheduling/cluster_resource_data.h

```diff
@@ -27,139 +27,546 @@
+      : ResourceRequest(resource_map, false){};
 
-/// Convert a vector of resource units to a vector of doubles.
-std::vector<double> VectorFixedPointToVectorDouble(
-    const std::vector<FixedPoint> &vector_fp);
+  ResourceRequest(absl::flat_hash_map<ResourceID, FixedPoint> resource_map,
+                  bool requires_object_store_memory)
+      : predefined_resources_(PredefinedResourcesEnum_MAX, 0),
+        requires_object_store_memory_(requires_object_store_memory) {
+    for (auto entry : resource_map) {
+      Set(entry.first, entry.second);
+    }
+  }
 
-struct ResourceCapacity {
-  FixedPoint total;
-  FixedPoint available;
-  ResourceCapacity() {}
-  ResourceCapacity(FixedPoint &&_available, FixedPoint &&_total)
-      : total(_total), available(_available) {}
-};
+  bool RequiresObjectStoreMemory() const { return requires_object_store_memory_; }
 
-/// Capacities of each instance of a resource.
-struct ResourceInstanceCapacities {
-  std::vector<FixedPoint> total;
-  std::vector<FixedPoint> available;
-};
+  /// Get the value of a particular resource.
+  /// If the resource doesn't exist, return 0.
+  FixedPoint Get(ResourceID resource_id) const {
+    auto ptr = GetPointer(resource_id);
+    if (ptr == 0) {
+      return FixedPoint(0);
+    } else {
+      return *ptr;
+    }
+  }
 
-// Data structure specifying the capacity of each resource requested by a task.
-class ResourceRequest {
- public:
-  /// List of predefined resources required by the task.
-  std::vector<FixedPoint> predefined_resources;
-  /// List of custom resources required by the task.
-  absl::flat_hash_map<int64_t, FixedPoint> custom_resources;
+  /// Set a resource to the given value.
+  /// NOTE: if the new value is 0, the resource will be removed.
+  ResourceRequest &Set(ResourceID resource_id, FixedPoint value) {
+    if (value == 0) {
+      if (IsPredefinedResource(resource_id)) {
+        predefined_resources_[resource_id.ToInt()] = 0;
+      } else {
+        custom_resources_.erase(resource_id.ToInt());
+      }
+    } else {
+      auto ptr = GetPointer(resource_id);
+      if (ptr == nullptr) {
+        custom_resources_[resource_id.ToInt()] = value;
+      } else {
+        *ptr = value;
+      }
+    }
+    return *this;
+  }
+
+  /// Check whether a particular resource exist.
+  bool Has(ResourceID resource_id) const {
+    auto ptr = GetPointer(resource_id);
+    return ptr != nullptr && *ptr != 0;
+  }
+
+  /// Clear the whole set.
+  void Clear() {
+    for (size_t i = 0; i < predefined_resources_.size(); i++) {
+      predefined_resources_[i] = 0;
+    }
+    custom_resources_.clear();
+  }
+
+  /// Cap the resource values in this set by those in another set.
+  void Cap(const ResourceRequest &other) {
+    for (auto &resource_id : ResourceIds()) {
+      auto this_value = Get(resource_id);
+      auto other_value = other.Get(resource_id);
+      if (this_value > other_value) {
```

Comment:
That makes sense. But in practice, the 2 sets can have different resources. Because the value in the available can be reduced to 0. 
Also, I think this method should be removed. If even "available > total", we can still keep that. I didn't want to change the behavior and then added this "Cap" method. I can do some investigation afterwards. 
For this PR, I'll make this method generic.